mix/CodeModel.cpp

```diff
@@ -327,6 +331,48 @@ void CodeModel::runCompilationJob(int _jobId)
 	emit stateChanged();
 }
 
+void CodeModel::gasEstimation(solidity::CompilerStack const& _cs)
+{
+	m_gasCostsMaps.clear();
+	for (std::string n: _cs.getContractNames())
+	{
+		ContractDefinition const& contractDefinition = _cs.getContractDefinition(n);
+		QString sourceName = QString::fromStdString(*contractDefinition.getLocation().sourceName);
+
+		if (!m_gasCostsMaps.contains(sourceName))
+			m_gasCostsMaps.insert(sourceName, QVariantList());
+
+		if (!contractDefinition.isFullyImplemented())
+			continue;
+		dev::solidity::SourceUnit const& sourceUnit = _cs.getAST(*contractDefinition.getLocation().sourceName);
+		AssemblyItems const* items = _cs.getRuntimeAssemblyItems(n);
+
+		StructuralGasEstimator estimator;
+		std::map<ASTNode const*, GasMeter::GasConsumption> gasCosts = estimator.breakToStatementLevel(estimator.performEstimation(*items, std::vector<ASTNode const*>({&sourceUnit})), {&sourceUnit});
+
+		for (auto gasIte = gasCosts.begin(); gasIte != gasCosts.end(); ++gasIte)
+		{
+			SourceLocation location = gasIte->first->getLocation();
+			GasMeter::GasConsumption cost = gasIte->second;
+			std::stringstream v;
+			v << cost.value;
+			GasMap* gas = new GasMap(location.start, location.end, QString::fromStdString(v.str()), gasIte->second.isInfinite);
+			m_gasCostsMaps.find(sourceName).value().push_back(QVariant::fromValue(gas));
+		}
+	}
+}
+
+QVariantList CodeModel::gasCostByDocumentId(QString const& _documentId) const
+{
+	if (m_gasCostsMaps.contains(_documentId))
```

Comment:
You could have just called find() without contains(), and then check for sourceMapIter != m_gasgCostsMaps.end()